django/chat/llm.py

```diff
@@ -0,0 +1,252 @@
+    def __init__(
+        self, deployment: str = settings.DEFAULT_CHAT_MODEL, temperature: float = 0.1
+    ):
+        if deployment not in self._deployment_to_model_mapping:
+            raise ValueError(f"Invalid deployment: {deployment}")
+        self.deployment = deployment
+        self.model = self._deployment_to_model_mapping[deployment]
+        self.temperature = temperature
+        self._token_counter = TokenCountingHandler(
+            tokenizer=tiktoken.encoding_for_model(self.model).encode
+        )
+        self._callback_manager = CallbackManager([self._token_counter])
+        self.llm = self._get_llm()
+        self.embed_model = self._get_embed_model()
+        self._service_context = ServiceContext.from_defaults(
+            llm=self.llm,
+            embed_model=self.embed_model,
+            callback_manager=self._callback_manager,
+        )
+        self.max_input_tokens = self._deployment_to_max_input_tokens_mapping[deployment]
+
+    # Convenience methods to interact with LLM
+    # Each will return a complete response (not single tokens)
+    async def chat_stream(self, chat_history: list):
+        """
+        Stream complete response (not single tokens) from list of chat history objects
+        """
+        response_stream = await self.llm.astream_chat(chat_history)
+        async for chunk in response_stream:
+            yield chunk.message.content
+
+    async def stream(self, prompt: str):
+        """
+        Stream complete response (not single tokens) from single prompt string
+        """
+        response_stream = await self.llm.astream_complete(prompt)
+        async for chunk in response_stream:
+            yield chunk.text
+
+    def complete(self, prompt: str):
+        """
+        Return complete response string from single prompt string (no streaming)
+        """
+        return self.llm.complete(prompt).text
+
+    async def tree_summarize(
+        self,
+        context: str,
+        query: str = "summarize the text",
+        template: PromptTemplate = None,
+    ):
+        """
+        Stream complete response (not single tokens) from context string and query.
+        Optional: summary template (must include "{context_str}" and "{query_str}".)
+        """
+        response = await self._get_tree_summarizer(
+            summary_template=template
+        ).aget_response(query, [context])
+        response_text = ""
+        async for chunk in response:
+            response_text += chunk
+            yield response_text
+
+    # Token counting / cost tracking
+    @property
+    def input_token_count(self):
+        return self._token_counter.prompt_llm_token_count
+
+    @property
+    def output_token_count(self):
+        return self._token_counter.completion_llm_token_count
+
+    @property
+    def embed_token_count(self):
+        return self._token_counter.total_embedding_token_count
+
+    def create_costs(
+        self, user: User, feature: str, embed_task: str = "query"
```

Comment:
to refactor: get rid of "embed_task". instead, make a feature "librarian".
| | Current | Proposed |
|--------|--------|--------|
| **feature** | qa | librarian |
| **cost_type** | embed-doc | embed |

| | Current | Proposed |
|--------|--------|--------|
| **feature** | qa | qa |
| **cost_type** | embed-query | embed |